src/fluree/db/merge.cljc

```diff
@@ -0,0 +1,262 @@
+                                                      source-branch-info target-branch-info))]
+
+      (when preview?
+        (reduced
+         {:status :preview
+          :operation :merge
+          :from from
+          :to to
+          :can-fast-forward can-ff?
+          :strategy (cond
+                      (and can-ff? (not= ff :never)) "fast-forward"
+                      squash? "squash"
+                      :else "merge")}))
+
+      (cond
+          ;; Fast-forward only mode - fail if not possible
+        (and (= ff :only) (not can-ff?))
+        {:status :error
+         :operation :merge
+         :from from
+         :to to
+         :error :db/cannot-fast-forward
+         :message "Fast-forward not possible - branches have diverged"}
+
+          ;; Fast-forward when possible (unless explicitly disabled)
+        (and can-ff? (not= ff :never))
+        (<? (ops/fast-forward! conn from to))
+
+          ;; Squash mode
+        squash?
+        (<? (ops/squash! conn from to opts))
+
+          ;; Regular merge mode (not yet implemented)
+        :else
+        (throw (ex-info "Regular merge not yet implemented. Use :squash? true or :ff :auto"
+                        {:status 501 :error :not-implemented}))))))
+
+(defn- recreate-branch
+  "Deletes and recreates a branch pointing to a new commit."
+  [conn branch-spec temp-branch new-commit-id branch-metadata]
+  (go-try
+    ;; Delete the existing branch
+    (<? (api.branch/delete-branch! conn branch-spec))
+    ;; Recreate it pointing to the new commit
+    (<? (api.branch/create-branch!
+         conn branch-spec temp-branch
+         (assoc branch-metadata :from-commit new-commit-id)))
+    ;; Ensure subsequent loads see fresh state
+    (connection/release-ledger conn branch-spec)))
+
+(defn rebase!
+  "Rebases source branch onto target branch (updates source branch).
+  
+  Takes the source branch and replays its commits on top of the target branch.
+  The source branch is updated with new commits, target branch remains unchanged.
+  
+  Parameters:
+    conn - Connection object
+    from - Source branch to rebase (will be updated)
+    to - Target branch to rebase onto (unchanged)
+    opts - Rebase options:
+      :squash? - Combine all commits into one (default false)
+      :preview? - Dry run without changes (default false)
+      
+  Returns promise resolving to rebase result."
+  [conn from to opts]
+  (go-try
+    (merge-branch/validate-same-ledger! from to)
+    (let [{:keys [squash? preview?]
+           :or {squash? false}} opts
+
+          ;; Load branches and find LCA
+          {:keys [source-db target-db source-branch-info target-branch-info]}
+          (<? (merge-branch/load-branches conn from to))
+
+          lca (<? (merge-branch/find-lca conn source-db target-db
+                                         source-branch-info target-branch-info))]
+
+      (when preview?
```

Comment:
This does not short circuit, so the rebasing code below will always execute whether or not the `preview?` option is set.